firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -198,4 +200,71 @@ void shutdown() {
       Thread.currentThread().interrupt();
     }
   }
+
+  /**
+   * Cancels the connectivityStateTimer if the new state indicates gRPC is online. Starts the timer
+   * if the new state indicates gRPC is connecting.
+   */
+  private void onConnectivityStateChange(ManagedChannel channel) {
+    ConnectivityState newState = channel.getState(false);
+    // Check that the new state is online, then cancel timer.
+    if (newState == ConnectivityState.READY) {
+      clearConnectivityTimer();
+    } else if (newState == ConnectivityState.CONNECTING) {
+      connectivityAttemptTimer =
+          asyncQueue.enqueueAfterDelay(
+              TimerId.CONNECTIVITY_ATTEMPT_TIMER,
+              CONNECTIVITY_ATTEMPT_TIMEOUT_MS,
+              () -> {
+                clearConnectivityTimer();
+                resetChannel(channel);
+              });
+    }
+    // Re-listen for next state change.
+    channel.notifyWhenStateChanged(
+        newState, () -> asyncQueue.enqueueAndForget(() -> onConnectivityStateChange(channel)));
+  }
+
+  /**
+   * Shuts down and re-initializes the channel.
+   *
+   * <p>This is used when the connectivity attempt timer elapses and we need to reset the gRPC
+   * channel to re-establish connectivity.
+   */
+  private void resetChannel(ManagedChannel channel) {
+    asyncQueue.enqueueAndForget(
+        () -> {
+          channel.shutdownNow();
+          initChannelTask();
+        });
+  }
+
+  private void initChannelTask() {
+    // We execute network initialization on a separate thread to not block operations that depend on
+    // the AsyncQueue.
```

Comment:
One potential concern with recreating the channel is it could be expensive.  We're already doing it on a background thread for performance reasons, so it's possible this adds some meaningful delay.

As part of your logging, I'd recommend logging the start/finish of this so that we can get timing data and see how long this takes each time.

Reply:
It's taking around 15-40ms to reset the channel from what I've seen based on the logs.

Reply:
Hrm.  That is longer than I hoped.  I don't know if it's a problem or not.  I suggest getting input from Gil and gRPC folks (and perhaps point out this delay to gRPC and see if they have other suggestions for implementing a connection timeout).

Reply:
👍